core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -132,9 +141,17 @@ class ReplicaFetcherThread(name: String,
 
     brokerTopicStats.updateReplicationBytesIn(records.sizeInBytes)
 
+    logAppendInfo.foreach { _ => partitionsWithNewRecords += topicPartition }
```

Comment:
Note that the HWM is updated separately here: `updateHighWatermarkAndStartOffset`.

Reply:
I am actually not sure about this. I need to think about it a little more. Let me know what you think.

Reply:
i think this makes sense. consumer fetches will at most fetch up to the HWM so we can add the topic partition only when `LeaderHwChange.Increased`

Reply:
Yeah, that's right. We just need to ensure that the HWM is always <= log end offset. Otherwise, this does not work. I will read this code again tomorrow.

Reply:
i believe that is the case. from [UnifiedLog.updatedHighWatermarkWithLogEndOffset()](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/log/UnifiedLog.scala#L597-L603)
```
  private def updateHighWatermarkWithLogEndOffset(): Unit = {
    // Update the high watermark in case it has gotten ahead of the log end offset following a truncation
    // or if a new segment has been rolled and the offset metadata needs to be updated.
    if (highWatermark >= localLog.logEndOffset) {
      updateHighWatermarkMetadata(localLog.logEndOffsetMetadata)
    }
  }
```
this is called on every `UnifiedLog.append()` and `UnifiedLog.roll()`

however, i did notice that `LeaderHWChange` is not set anywhere else except in the produce code path. from `Partition.appendRecordsToLeader()`:
```
info.copy(leaderHwChange = if (leaderHWIncremented) LeaderHwChange.Increased else LeaderHwChange.Same)
```
no other `LogAppendInfo` initialization passes in LeaderHWChange.

so, the follower would need to keep track of whether the high watermark changed for each partition which i don't think is something we want to do.

Reply:
I am still not fully convinced here. One aspect to remember is that the HWM is updated asynchronously and that the fetch request is bound by the HWM, not the log end offset. Let's imagine the following events:
1. The log is empty.
2. The consumer fetches from the follower at offset 0. The fetch request goes to the purgatory because the requested offset is not available yet.
3. The producer produces one request to the leader.
4. The follower fetches the new records and happens them to its local log. The log end offsets moves to offset 0 but the HWM stays at 0 for now. This trigger the purgatory because new records got appended.
5. The delayed fetch checks if the fetch request can be completed. It should not at this point because the HWM is still at offset 0.
6. The follower fetches again to commit the new records and to get the new HWM. It updates the HWM locally to offset 1. If there are no new records, we don't trigger the purgatory now. The delayed fetch remains in the purgatory until the timeout is reached.

This seems to be a legitimate course of actions, isn't it? Moreover, there is a race condition between 4 and 5 so the test may pass sometimes.

Reply:
Good point, David!

Reply:
> 4. The follower fetches the new records and happens them to its local log. The log end offsets moves to offset 0 but the HWM stays at 0 for now. This trigger the purgatory because new records got appended.

log end offset moves to offset 1 right?

i don't think there can be a race at step 5 & 6 (and not 4 & 5 right) because it's the same replica fetcher thread that updates the follower's HWM then completes the purgatory. that said, the current test has been working successfully consistently because
`logAppendInfo` in `ReplicaFetcherThread.processPartitionData()` 
```
    // Append the leader's messages to the log
    val logAppendInfo = partition.appendRecordsToFollowerOrFutureReplica(records, isFuture = false)
```
is not None. it's in the form of
```
LogAppendInfo(None,-1,None,-1,-1,-1,0,RecordConversionStats(temporaryMemoryBytes=0, numRecordsConverted=0, conversionTimeNanos=0),NoCompressionCodec,NoCompressionCodec,0,0,true,-1,List(),null,None)
```
so the follower was completing the purgatory both when it appended the new record, and again when it updated the HWM.

Reply:
@dajac i think in order to resolve the scenario you've mentioned, we would ideally only want to complete the purgatory when HWM is updated. and since HWM update is not asynchronous for the follower, it seems safe. what do you think?